### ISSUE TITLE
Feature: Create a runtime diagram from a code snapshot

### ISSUE THREAD

**Issue by Dedekind561** [CodeYourFuture/curriculum#8]
A super useful thing to have would be a means of generating a runtime diagram from a snippet of code.
For example, given a code block like this:
```js
function sum(a,b) {
      return a + b;
}
const result = sum(a,b);
```
I want to indicate a specific line in this piece of code and from there take a snapshot of the following things:
Call stack, memory, thread
This would almost be like taking a snapshot of the debugger output - but formatted in a nice diagram.

<img width="691" alt="image" src="https://github.com/CodeYourFuture/curriculum/assets/25401570/ffe78df2-8483-4572-bb16-b5173d3c0cc6">

**Comment by Dedekind561** [comment]
@gregdyke I think what you're working on at the moment would help with this issue

**Comment by gregdyke** [comment]
yep. We should call and see exactly what you'd like to see (as technically I can do it, just want the detail)

**Comment by SallyMcGrath** [comment]
It feels like if your thing is using Acorn and we can get an AST from that, we could translate that into Mermaid, as we're already using Mermaid?

Mermaid generates SVGs from markup using D3, basically, so although ideally we'd use a prebuilt one, it doesn't seem insane to build our own _if we find we need to_, [just painstaking](https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/diagrams/timeline/timeline.spec.js). (It will be laborious)

Do take a proper look through and try with the existing stuff (maybe state or C4) before we go rogue... 🙏 😉

**Comment by Dedekind561** [comment]
Yeah that example was very poor 😂 
Here is a better one:
<img width="925" alt="image" src="https://github.com/CodeYourFuture/curriculum/assets/25401570/c5407524-3b97-4a37-b669-a7eecd74d4ee">

**Comment by Dedekind561** [comment]
@gregdyke here is my [appointment schedule](https://calendar.google.com/calendar/u/0/appointments/schedules/AcZssZ2NT2uO8RV_aF24bVSAH3tmAx3xHdn9je3N19cOPzMuqV1Okl6oymS6F8n1vxAiu-cm4XDjPLHn)

**Comment by Dedekind561** [comment]
> Yeah that example was very poor 😂 Here is a better one: <img alt="image" width="925" src="https://user-images.githubusercontent.com/25401570/239910162-c5407524-3b97-4a37-b669-a7eecd74d4ee.png">

**Comment by SallyMcGrath** [comment]
> Yeah, that part is "easy" (though suggestions of existing implementations to re-use/adapt are welcome). But in @Dedekind561 example, I don't actually see a "diagram" (or at least not the kind I would have in mind from the title)

I very carefully did not say easy 😛

**Comment by Dedekind561** [comment]
😂 "easy" is now on record, folks.

**Comment by Dedekind561** [comment]
@gregdyke could you also build this too please: https://py.mit.edu/spring23/practice/diagram1 😂

**Comment by Dedekind561** [comment]
Actually it's all open source 😎

**Comment by Dedekind561** [comment]
This is also similar to what you're building: https://pythontutor.com/visualize.html#mode=display
